libs/api/utils/src/lib/api-utils.ts

```diff
@@ -1,8 +1,31 @@
+import { prisma } from 'apps/coursition/prisma/prismaClient'
+import { parseApiKey } from '../core/api/parseApiKey'
+import { isDateBeforeNow } from '../core/date/isDateBeforeNow'
+import type { ApiErrorCode } from '../errorList'
+
 export function reportUsage(apiKey: string, duration: number, type: 'video' | 'document' | 'web') {
   console.log(`API Key ${apiKey} used ${duration} on ${type}.`)
 }
 
-export function validateApiKey(apiKey: string) {
-  if (!apiKey) return false
-  return true
+export async function validateApiKey(apiKey: string): Promise<ApiErrorCode | undefined> {
+  if (!apiKey) return 'PUBLIC_API_KEY_DOES_NOT_EXIST'
+  if (!parseApiKey(apiKey).isValid) return 'PUBLIC_API_KEY_IS_NOT_IN_VALID_FORMAT'
+
+  const key = await prisma.cas__organisation_api_key.findFirst({
+    select: { id: true, organisation_id: true, is_active: true, is_deleted: true, expiration_date: true },
+    where: { api_key: apiKey },
+  })
+  if (!key) return 'PUBLIC_API_KEY_DOES_NOT_EXIST'
+  if (key.is_deleted) return 'PUBLIC_API_KEY_HAS_BEEN_DELETED'
+  if (!key.is_active) return 'PUBLIC_API_KEY_IS_NOT_ACTIVE'
+  if (isDateBeforeNow(key.expiration_date)) return 'PUBLIC_API_KEY_HAS_BEEN_EXPIRED'
+
+  await incrementKeyUsage(key.id)
+}
+
+async function incrementKeyUsage(keyId: number) {
```

Comment:
**question:** How is this related to validation of API key? Also do we want to count every validation as "using API key" and if so what are the reasons for that? :)

Reply:
It is a preparation for future use. If we count all requests with this API key, and then the consumption of individual requests, we can easily discover in the future what percentage of requests we cannot count correctly.

At the same time, we will know which keys are used and which are not. And we can also store the date of the last use of the key. I would recommend letting inactive keys expire later.

At the same time, if a key has been used many times (hundreds of thousands of requests per hour), it may have been stolen and is being misused by a scraping robot.

We may then reveal other usage statistics that may be useful to administrators.

Reply:
Awesome, I see your experience is invaluable as it would have taken me ton of time to realize this! Closing.